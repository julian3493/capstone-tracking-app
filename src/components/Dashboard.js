import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import axios from 'axios';
import Logout from './Logout';
import Navigation from './Navigation';
import TopBar from './TopBar';
import '../styles/Dashboard.css';

const devices = [
  'pc_laptop',
  'tv',
  'videogames_console',
  'blueray',
  'pc_desktop',
];

const setOpt = () => devices.map((device) => <option value={device} key={device}>{device}</option>);

const Dashboard = () => {
  const user = useSelector((state) => (state.logged_user));

  const [track, setTrack] = useState({
    electrodomestic: 'pc_laptop',
    day: '',
    time_connected: 0,
    watts: 0,
  });

  const fetchData = async () => {
    await axios.post('http://localhost:3001/tracks', {
      track: {
        electrodomestic: track.electrodomestic,
        day: track.day,
        time_connected: track.time_connected,
        watts: track.watts,
        user_id: user.id,
      },
    },
    { withCredentials: true }).then((response) => {
      console.log(response);
    }).catch((err) => {
      throw (err);
    });
  };

  const handleSubmit = (e) => {
    fetchData();
    e.preventDefault();
  };

  const handleChange = (e) => {
    setTrack({
      ...track,
      [e.target.name]: e.target.value,
    });
    e.preventDefault();
  };

  return (
    <div className="screen">
      {TopBar({ title: 'Dashboard' })}
      <form className="track-form" onSubmit={handleSubmit}>
        <p className="select-label">Select Type:&nbsp;&nbsp;</p>
        <select name="electrodomestic" onChange={handleChange}>
          {setOpt()}
        </select>
        <input type="date" name="day" value={track.day} onChange={handleChange} required />
        <input type="number" name="time_connected" placeholder="Time" value={track.time_connected} onChange={handleChange} required />
        <input type="number" name="watts" placeholder="Watts" value={track.watts} onChange={handleChange} required />
        <button type="submit">Create</button>
      </form>
      <Logout />
      <Navigation />
    </div>
  );
};

export default Dashboard;
