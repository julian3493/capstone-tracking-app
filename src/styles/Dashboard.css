.screen {
  display: flex;
  flex-direction: column;
  height: 100vh;
}

.track-form {
  margin: 40px 35px;
  padding: 20px 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  background-color: #f3f4f6;
}

.select-label {
  margin-bottom: 5px;
  color: #41b5e8;
}

.track-form input,
.track-form select {
  margin-bottom: 10px;
  height: 30px;
}
