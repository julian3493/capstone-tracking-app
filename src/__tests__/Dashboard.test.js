import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import Dashboard from '../components/Dashboard';
import store from '../redux/store';

it('it tests the render Dashboard correctly', () => {
  const dashboard = renderer
    .create(<Provider store={store}>
      <BrowserRouter>
        <Dashboard />
      </BrowserRouter>
    </Provider>)
    .toJSON();
  expect(dashboard).toMatchSnapshot();
});

describe('render correctly the topbar component', () => {
  it('renders the topbar', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Dashboard />
        </BrowserRouter>
      </Provider>,
    );
    const topbar = document.querySelector('.top');
    expect(topbar).toBeDefined();
  });

  it('renders the page title', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Dashboard />
        </BrowserRouter>
      </Provider>,
    );
    const topbar = document.querySelector('.top');
    expect(topbar.firstChild.innerHTML).toEqual('Dashboard');
  });
});

describe('render correctly the dashboard content', () => {
  it('should display a form', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Dashboard />
        </BrowserRouter>
      </Provider>,
    );
    const form = document.querySelector('.track-form');
    expect(form).toBeDefined();
  });

  it('should have a select input', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Dashboard />
        </BrowserRouter>
      </Provider>,
    );
    const form = document.querySelector('.track-form');
    const select = document.querySelector('select');
    expect(form.children[1]).toEqual(select);
  });

  it('should have a date input', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Dashboard />
        </BrowserRouter>
      </Provider>,
    );
    const form = document.querySelector('.track-form');
    expect(form.children[2].type).toEqual('date');
  });

  it('should have a number input for time', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Dashboard />
        </BrowserRouter>
      </Provider>,
    );
    const form = document.querySelector('.track-form');
    expect(form.children[3].type).toEqual('number');
  });

  it('should have a number input for watts', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Dashboard />
        </BrowserRouter>
      </Provider>,
    );
    const form = document.querySelector('.track-form');
    expect(form.children[4].type).toEqual('number');
  });
});

it('should render a logout button', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Dashboard />
      </BrowserRouter>
    </Provider>,
  );
  const logoutBtn = document.querySelector('.logoutBtn');
  expect(logoutBtn).toBeDefined();
});

it('should render a navigation component', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <Dashboard />
      </BrowserRouter>
    </Provider>,
  );
  const nav = document.querySelector('.nav-div');
  expect(nav).toBeDefined();
});