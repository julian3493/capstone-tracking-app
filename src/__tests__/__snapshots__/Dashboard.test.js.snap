// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it tests the render Dashboard correctly 1`] = `
<div
  className="screen"
>
  <div
    className="top"
  >
    <p
      className="top-title"
    >
      Dashboard
    </p>
  </div>
  <form
    className="track-form"
    onSubmit={[Function]}
  >
    <p
      className="select-label"
    >
      Select Type:  
    </p>
    <select
      name="electrodomestic"
      onChange={[Function]}
    >
      <option
        value="pc_laptop"
      >
        pc_laptop
      </option>
      <option
        value="tv"
      >
        tv
      </option>
      <option
        value="videogames_console"
      >
        videogames_console
      </option>
      <option
        value="blueray"
      >
        blueray
      </option>
      <option
        value="pc_desktop"
      >
        pc_desktop
      </option>
    </select>
    <input
      name="day"
      onChange={[Function]}
      required={true}
      type="date"
      value=""
    />
    <input
      name="time_connected"
      onChange={[Function]}
      placeholder="Time"
      required={true}
      type="number"
      value={0}
    />
    <input
      name="watts"
      onChange={[Function]}
      placeholder="Watts"
      required={true}
      type="number"
      value={0}
    />
    <button
      type="submit"
    >
      Create
    </button>
  </form>
  <button
    className="logoutBtn"
    onClick={[Function]}
    type="button"
  >
    Logout
  </button>
  <div
    className="nav-div"
  >
    <a
      className="nav-link"
      href="/dashboard"
      onClick={[Function]}
    >
      Dashboard
    </a>
    <a
      className="nav-link"
      href="/tracks"
      onClick={[Function]}
    >
      Tracks
    </a>
    <a
      className="nav-link"
      href="/statistics"
      onClick={[Function]}
    >
      Statistics
    </a>
    <a
      className="nav-link"
      href="/more"
      onClick={[Function]}
    >
      More
    </a>
  </div>
</div>
`;
